Service timeout

mattcenoo opened the pull request:
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (#1248 ) |
| Primary OS tested on | (Ubuntu 18.04) |
| Robotic platform tested on | (Yunji Robot) |

---

## Description of contribution in a few bullet points

* Add a parameter to control the timeout of every specific action/service
``` <ClearEntireCostmap service_name="/local_costmap/clear_entirely_local_costmap" server_timeout="200"/> ```
* Add  a warning  stating that the service timed out 

---

## Future work that may be required in bullet points
* There might be another elegant way to fix this problem, cz' clear a 10000x10000pix costmap  may take a while.
<!--
* I think there might be some optimizations to be made from STL vector
* I see alot of redundancy in this package, we might want to add a function `bool XYZ()` to reduce clutter
* I tested on a differential drive robot, but there might be issues turning near corners on an omnidirectional platform
-->

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp:
```diff
@@ -45,8 +45,9 @@ class BtServiceNode : public BT::CoroActionNode
-      "please_set_service_name_in_BT_Node"}};
+    static BT::NodeParameters params = {
+      {"service_name", "please_set_service_name_in_BT_Node"},
+      {"service_timeout", "10"}};
```
It seems like this is the same as `node_loop_timeout_` (effectively since all that's happening in this node is calling a service). Why not just use the `node_loop_timeout_` variable instead of introducing a new one?

mattcenoo replied:
Actually, I did hesitate about this variable, and thought natural to write as ```<SomeService service_timeout="100"/> ```. Btw,  using the node_loop_timeout_ variable may get confused with the blackboard variable node_loop_timeout. But this is not critical, I'll do that since you think node_loop_timeout_ is better.

SteveMacenski replied:
@crdelsey @mkhansen-intel  what do you think?

My understanding from reading the code is that BT navigator sets the default node loop timeout to 10ms to the blackboard. Then the service/action node base classes read from the blackboard this 10ms. That indicates to me that @mkhansen-intel's suggestion in the ticket to override that variable in the BT tag wouldn't work if I understand the difference between the blackboard an tag parameters correctly. 

So if node loop timeout is set by BT navigator, and then stored in the base classes, the only use of that variable that I see is actually in the timeouts to the service/action servers, not having anything to do with the timeout for the `node` itself. So we can just keep that variable.

If I'm missing another use of that variable `node_loop_timeout` to timeout actual behavior tree nodes, correct me. Or If I'm missing that the blackboard and the registered tag parameters are somehow connected to overload, also correct me.

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp:
```diff
@@ -59,10 +60,6 @@ class BtServiceNode : public BT::CoroActionNode
   {
     node_ = blackboard()->template get<rclcpp::Node::SharedPtr>("node");
 
-    // Get the required items from the blackboard
```
This is a necessary variable, please readd

mattcenoo replied:
done

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp:
```diff
@@ -86,12 +83,13 @@ class BtServiceNode : public BT::CoroActionNode
+        future_result, std::chrono::milliseconds(service_timeout_));
+    if (rc == rclcpp::executor::FutureReturnCode::SUCCESS) {
       return BT::NodeStatus::SUCCESS;
+    } else if (rc == rclcpp::executor::FutureReturnCode::TIMEOUT) {
```
Please add a `RCLCPP_WARN(node_->get_logger(), "Node timed out while executing service call to %s.", service_name_.c_str())'`

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/clear_costmap_service.hpp:
```diff
@@ -34,6 +34,7 @@ class ClearEntireCostmapService : public BtServiceNode<nav2_msgs::srv::ClearEnti
   : BtServiceNode<nav2_msgs::srv::ClearEntireCostmap>(service_node_name, params)
   {
     getParam<std::string>("service_name", service_name_);
+    getParam<unsigned int>("service_timeout", service_timeout_);
```
I  think all of these service timeouts are the same of the node loop timeout

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp:
```diff
@@ -105,6 +103,14 @@ class BtServiceNode : public BT::CoroActionNode
+  // a timeout waiting for a result that hasn't been received yet
+  virtual void on_service_timeout()
+  {
+    RCLCPP_WARN(node_->get_logger(), "\"%s\" service timed out with tolerance: %dms",
```
This virtual should be empty. If someone overrides it to use the function, we'd still like this warning to appear, see commment above

mattcenoo replied:
Right, my mistake.

SteveMacenski commented:
Good stuff!

So we use the variable node_loop_timeout_ across all our node base classes (https://github.com/ros-planning/navigation2/search?q=node_loop_timeout_&unscoped_q=node_loop_timeout_) so we can control the timeout of the service, I think you can just use that variable as is rather than adding a new one. 

Then just change the BT to get that variable and you should be good to go (with a log statement)!

mattcenoo commented:
I'm getting kinda confused, let me make it clear.
1. delete the _service_timeout_ variable;
2. add an action _SetBlackboard_ in the tree, set _node_loop_timeout_ to 1000ms（_clear_costmap_ is the long-board on the barrel), then all actions and services share this timeout; (add another _action_timeout_ variable seems more flexible w.r.t. _node_loop_timeout_ :) )
Please confirm this, thanks.

SteveMacenski commented:
I'd say
1) I think `server_timeout` (server rather than service) is fine to rename `node_loop_timeout` since I think it more accurately describes what's happening. I'm going to actually reverse what I said before I think setting it as a parameter to be defined as an xml tag makes more sense than what we currently did as a blackboard parameter. This is because this parameter should be really defined for each service or action depending on how long it should run, however it should *default* on construction to the blackboard parameter set by BT navigator when the parameter is not defined in the behavior tree xml file. 

EI in the `on_init()` method should be 
```
server_timeout_ = blackboard()->template get<std::chrono::milliseconds>("server_timeout");
getParam<std::string>("server_timeout", server_timeout_);
```
As to default to the blackboard specified parameter in the absence of the xml provided version.

 @mkhansen-intel @crdelsey what are your thoughts? 

2) Right, you have the same thought as me the action node should be the same style. That's why I prefer if you actually changed the name from `service_timeout` to `server_timeout` since both `service`  and `action` are `servers` so that one parameter name would work in both of these base classes. 

Hold off on the Blackboard vs xml tag parameters for the moment until we get feedback from the others, I don't want to make a design change, have you do it, and have them disagree then make you do it over again.

mattcenoo commented:
It's ok, I agree with your opinions. I'll wait for you guys' final decision.

crdelsey commented:
> I now agree with you that it should be a node-based parameter rather than only a blackboard parameter, so that for each individual call you can specify the timeout if you choose

With version 3 of the behavior tree (V3), I don't think we need to make the distinction. If we write:
```
<ClearEntireCostmap service_timeout="200"/>
```
we set this particular element's timeout to 200. If we write

```
<ClearEntireCostmap service_timeout="{service_timeout}"/>
```

we grab it from the blackboard port called service_timeout.

If we can get `service_timeout="{service_timeout}"` to be default, we'd grab the timeout off the blackboard, and we can override it at any time in the XML.

I need to do some experimentation with default values still. I'll see if I can get this to work in V3 tonight.

crdelsey commented:
As I re-read this, maybe I'm misinterpreting the question.

If the question is just, should we make the server_timeout be something that can be specified in the XML, I've got no objections.

The behavior tree docs caution against too many ports in the XML, 

https://github.com/BehaviorTree/BehaviorTree.CPP/blob/ver_3/docs/tutorial_08_additional_args.md#custom-initialization-andor-construction

but I don't have a good feel for what is too many yet. If we feel there is a need for different timeouts for different services, then, making it a XML parameter seems the right thing to do. And I think having different timeouts makes sense.

I'd say, go ahead with that change.

SteveMacenski commented:
I'm hearing agreement. @mattcenoo please make the changes requested.

@crdelsey on 
> If we can get service_timeout="{service_timeout}" to be default, we'd grab the timeout off the blackboard, and we can override it at any time in the XML.

Well the point I was trying to get across is what happens when someone fails to specify anything at all. If we do nothing then the timeout will be `0` and that's unacceptable, so that's why I ask for a call to the blackboard then override so that either way a valid default value is taken. It is always good to be explicit and do as you suggest above if not setting to override but not strictly necessary.

crdelsey commented:
> If we do nothing then the timeout will be 0 and that's unacceptable, so that's why I ask for a call to the blackboard then override so that either way a valid default value is taken.

That seems reasonable for now. When I can get default values to work in the new version of the library, we can then remove the extra call to read from the blackboard.

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -106,9 +115,9 @@ class BtActionNode : public BT::CoroActionNode
     rclcpp::executor::FutureReturnCode rc;
     do {
-      rc = rclcpp::spin_until_future_complete(node_, future_result, node_loop_timeout_);
+      rc = rclcpp::spin_until_future_complete(node_, future_result, server_timeout_);
```
These changes don't reflect the same changes you have in the bt_service_node, make sure you're keeping these in sync

mattcenoo replied:
I've reverted _bt_action_node_ as you mentioned above.

SteveMacenski replied:
Sorry maybe I should have been more clear:

The BTActionNode and BTServiceNode are essentially the same thing with the same purpose but one calls an action server and the other calls the service server. I wanted them both to use the same variable set, the new `server_timeout_` & all the exact changes you made to the BTServiceNode were good, but I wanted them identically reflected in BTActionNode

mattcenoo replied:
Do you mean that I should use a variable like _action_server_timeout__ in _BTActionNode_ and a _service_server_timeout__ in _BTServiceNode_?  And keep using _node_loop_timeout_ in blackboard or change it to _server_timeout_?

SteveMacenski commented:
This is failing also all the integration level CI tests

SteveMacenski commented:
To be clear, line by line, all changes made to bt service node I want to bt action node. 

this is in addition to making CI pass

orduno commented:
@mattcenoo so this PR addresses the BT services side of #1070, you also want to do the BT actions to get the [full point](https://github.com/ros-planning/navigation2/milestone/16)?

mattcenoo commented:
@orduno Yes, that's what I'd done in commit af544c3.

@SteveMacenski If you mean that adding a timeout warning in bt action node like _bt_service_node_ do, then it's not appropriate I think, there will be warning-flood in the log because it's normal to be timed out for action ```FollowPathAction```. Any other changes are identical between  BTActionNode and BTServiceNode in commit af544c3 except this warning. If I'm missing anything, correct me.

SteveMacenski commented:
I'm not sure we're on the same page. As an exhaustive list, inside of BT service node you:
- renamed `node_loop_timeout_` for `server_timeout_`
- You did a `getInput` to get the input from the XML, if present
- You check for the blackboard `server_timeout` parameter beforehand
- Added the server_timeout port

I would like those identical changes to BT action node.

In addition before reverting you had:
-  renamed the blackboard parameter `node_loop_timeout` to `server_timeout` for the default
-  renamed the blackboard set for the node loop timeout in BT navigator to server_timeout

Please readd those. These BT base nodes should contain the same parameters of the same name with the same handling. The `if (rc != rclcpp::executor::FutureReturnCode::SUCCESS) {` to `==` and a couple else statements with a log will be different because the action server is expected to timeout (which @mkhansen-intel this is probably not good to be doing, we're even using what I consider to be the 'right toolset' already. We should be `while` looping over `get_status()` and when the status is `STATUS_SUCCEEDED` then we spin the future to get the result.)

mattcenoo commented:
>I'm not sure we're on the same page. As an exhaustive list, inside of BT service node you:
renamed node_loop_timeout_ for server_timeout_
You did a getInput to get the input from the XML, if present
You check for the blackboard server_timeout parameter beforehand
Added the server_timeout port
I would like those identical changes to BT action node.

Identical already.

>In addition before reverting you had:
renamed the blackboard parameter node_loop_timeout to server_timeout for the default
renamed the blackboard set for the node loop timeout in BT navigator to server_timeout
Please readd those

ok, no proplem.

>The if (rc != rclcpp::executor::FutureReturnCode::SUCCESS) { to == and a couple else statements with a log will be different because the action server is expected to timeout .

Not sure what to do, they have different expectation behavior as you mentioned, we expect service server to succeed while feeling no-surprise for the timed-out of action server.

SteveMacenski commented:
> Identical already.

I'm looking at the diff in the git view, I don't see any modifications to the BTActionNode, which would be required to at minimum change the parameter names and the BTNavigator to change the blackboard set.

> Not sure what to do, they have different expectation behavior as you mentioned, we expect service server to succeed while feeling no-surprise for the timed-out of action server.

Correct, you don't have to do anything. Just make sure that all the changes except for those in the timeout look from Service is the same in Action -- so doing the bulleted action items

mattcenoo commented:
>I'm looking at the diff in the git view, I don't see any modifications to the BTActionNode, which would be required to at minimum change the parameter names and the BTNavigator to change the blackboard set.

Sorry, I mean in commit af544c3

>Correct, you don't have to do anything. Just make sure that all the changes except for those in the timeout look from Service is the same in Action -- so doing the bulleted action items

I'll do that

SteveMacenski commented:
@mattcenoo lets get that commit to be in the PR again, check the linters `ament_cpplint` `ament_uncrustify` and lets get this PR in!

Sorry for the confusion, it is not this ordinarily complex, apologies.

mattcenoo commented:
Ah, sorry, I notice that:
```static BT::PortsList providedPorts()```
will be overridden by subclasses of BTActionNode and BTServiceNode.
Is there any other method to resolve this except to copy the ```BT::InputPort<std::chrono::milliseconds>("server_timeout")``` everywhere.

SteveMacenski commented:
Sounds good

SteveMacenski commented:
@crdelsey can you take another look? A bunch of stuff changed might worth relooking at

mattcenoo commented:
Yes, I've tested with the gazebo. So I have no idea of why the CI tests failed.

SteveMacenski commented:
Yeah the tests are busted right now. there's some work in the background for that. 

If you pull in master (or rebase) you'll see most of them go away. Can you actually do that for me? If some of them pass that tells me this is at least in concept working without me having to spin up my server

mattcenoo commented:
Done. Better now, one falling left (test_multi_robot).

SteveMacenski commented:
Awesome!
Thanks for your contribution! I really appreciate the work and getting through the comments
